File: paradygmaty/lista2/GF.java
public class GF {
    private int val;
    private int characteristic;

    public GF(int val, int characteristic) {
        this.characteristic = characteristic;
        this.val = val % characteristic;
    }

    public GF(int val) {
        this(val, 1234577);
    }

    @Override
    public String toString() {
        return String.format("%d, %d", val, characteristic);
    }

    private void checkErr(GF other) {
        if (this.characteristic != other.characteristic) {
            throw new IllegalArgumentException("Characteristics of GF elements do not match.");
        }
    }

    public boolean lessThan(GF other) {
        checkErr(other);
        return this.val < other.val;
    }

    public boolean greaterThan(GF other) {
        checkErr(other);
        return this.val > other.val;
    }

    public boolean lessThanOrEqual(GF other) {
        checkErr(other);
        return this.val <= other.val;
    }

    public boolean greaterThanOrEqual(GF other) {
        checkErr(other);
        return this.val >= other.val;
    }

    public boolean equals(GF other) {
        checkErr(other);
        return this.val == other.val;
    }

    public boolean notEquals(GF other) {
        checkErr(other);
        return this.val != other.val;
    }

    public GF add(GF other) {
        checkErr(other);
        return new GF((this.val + other.val) % this.characteristic, this.characteristic);
    }

    public GF subtract(GF other) {
        checkErr(other);
        return new GF((this.val + (other.characteristic - other.val)) % this.characteristic, this.characteristic);
    }

    public GF multiply(GF other) {
        checkErr(other);
        return new GF((this.val * other.val) % this.characteristic, this.characteristic);
    }

    public GF divide(GF other) {
        checkErr(other);
        int[] result = extendedGCD(other.val, other.characteristic);
        int gcd = result[0];
        int x = result[1];
        if(x < 0){
            x = other.characteristic - Math.abs(x);
        }
        if (gcd != 1) {
            throw new ArithmeticException("Division not possible.");
        }

        return new GF((this.val * (x % other.characteristic)) % this.characteristic, this.characteristic);
    }

    public void addInPlace(GF other) {
        checkErr(other);
        this.val = (this.val + other.val) % this.characteristic;
    }

    public void subtractInPlace(GF other) {
        checkErr(other);
        this.val = (this.val + (this.characteristic - other.val)) % this.characteristic;
    }

    public void multiplyInPlace(GF other) {
        checkErr(other);
        this.val = (this.val * other.val) % this.characteristic;
    }

    public void divideInPlace(GF other) {
        checkErr(other);
        int[] result = extendedGCD(other.val, other.characteristic);
        int gcd = result[0];
        int x = result[1];
        if(x < 0){
            x = other.characteristic - Math.abs(x);
        }
        if (gcd != 1) {
            throw new ArithmeticException("Division not possible.");
        }

        this.val = (this.val * (x % other.characteristic)) % this.characteristic;
    }

    private int[] extendedGCD(int a, int b) {
        if (b == 0) {
            return new int[] { a, 1, 0 };
        } else {
            int[] temp = extendedGCD(b, a % b);
            int d = temp[0];
            int x = temp[2];
            int y = temp[1] - (a / b) * temp[2];
            return new int[] { d, x, y };
        }
    }
}

File: paradygmaty/lista2/gf.py
class GF:
    def __init__(self, val: int, characteristic = 1234577 ):
        self.characteristic = characteristic
        self.val = val % characteristic

    def __str__(self):
        return str(f"{self.val}, {self.characteristic}")
    
    def __lt__(self, other):
        GF.__check_err(self, other)
        return self.val < other.val

    def __gt__(self, other):
        GF.__check_err(self, other)
        return self.val > other.val

    def __le__(self, other):
        GF.__check_err(self, other)
        return self.val <= other.val

    def __ge__(self, other):
        GF.__check_err(self, other)
        return self.val >= other.val

    def __eq__(self, other):
        GF.__check_err(self, other)
        return self.val == other.val

    def __ne__(self, other):
        GF.__check_err(self, other)
        return self.val != other.val

    def __add__(self, other):
        GF.__check_err(self, other)
        return GF( (self.val + other.val ) % self.characteristic, self.characteristic )

    def __sub__(self, other):
        GF.__check_err(self, other)
        return GF( (self.val + (-other.val % other.characteristic) ) % self.characteristic, self.characteristic )

    def __mul__(self, other):
        GF.__check_err(self, other)
        return GF( (self.val * other.val ) % self.characteristic, self.characteristic )

    def __truediv__(self, other):
        GF.__check_err(self, other)
        gcd, x, _ = GF.__euklides(other.val, other.characteristic)
        if(gcd != 1):
            return -1
        
        return GF( (self.val * (x % other.characteristic)) % self.characteristic, self.characteristic )

    def __euklides(a,b):
        if b == 0:
            return a, 1, 0
        else:
            d, x, y = GF.__euklides(b, a % b)
        return d, y, x - (a // b) * y
    
    def __check_err(c1, c2):
        if(c1.characteristic != c2.characteristic):
            raise ValueError